### PR TITLE
Behavior cloning state dim

### PR DESCRIPTION
**Pull request by jimzers**
Commands for reproducing:

1. Shape generation
2. Pretraining
3. Training

1:
```
python main_DDPGfD.py --saving_dir variable_speed_07132021_v01 --hand_orientation normal --shapes CubeM --with_grasp_reward False --with_orientation_noise False --max_episode 100 --controller_type position-dependent --mode position-dependent
```

2:
```
python main_DDPGfD.py --saving_dir state_dim_variable_speed_07132021_v02 --hand_orientation normal --shapes CubeM --with_orientation_noise False --expert_replay_file_path "./experiments/position-dependent/no_noise/no_grasp/" --with_grasp_reward False --max_episode 100 --controller_type policy --mode pre-train --update_after 30 --eval_num 30 --state_range all_real
```

3:
```
python main_DDPGfD.py --saving_dir training_07132021_v02 --hand_orientation normal --shapes CubeM --with_orientation_noise True --expert_replay_file_path "./experiments/position-dependent/no_noise/no_grasp/" --agent_replay_buffer_path "./experiments/pre-train/state_dim_variable_speed_07132021_v02/replay_buffer/" --pretrain_policy_path "./experiments/pre-train/state_dim_variable_speed_07132021_v02/policy/pre-train_DDPGfD_kinovaGrip" --max_episode 100 --controller_type policy --update_after 6 --eval_num 5 --mode train --state_range all_real
```

### CODE REVIEW

**Comment by jimzers** [comment]
tested on evaluation / training

File: gym-kinova-gripper/main_DDPGfD.py [hunk 1]
```diff
@@ -1352,6 +1371,107 @@ def create_input_variation_reward_plot(policies, start_episode, eval_freq, max_e
...
+    f1_f2_joint_angles_idx = np.array([27, 28, 30, 31])  # TODO: figure out the right angles here... => in order: f1 proximal angle joint, f2 proximal angle joint, f1 distal angle joint, f2 distal angle joint.
+    # note: these are also all in radians in simulator...
+
+    wrist_pos_idx = np.arange(18, 21)
+    obj_pos_idx = np.arange(21, 24)
+    joint_states_idx = np.arange(24, 33)
+    obj_size_idx = np.arange(33, 36)
+    finger_obj_dist_idx = np.arange(36, 48)
+
+    finger_obj_dist_f1_prox_1 = np.array([37])
```
Comment (nigelswenson) [inline gym-kinova-gripper/main_DDPGfD.py:1432]: incorrect. according to the comment above, this should be index 36. We should check all the indexes here and in kinova gripper env

File: gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py [hunk 2]
```diff
@@ -33,112 +32,150 @@
...
+import threading  # oh boy this might get messy
+from PIL import Image, ImageFont, ImageDraw  # Used to save images from rendering simulation
 import shutil
 
+import xml.etree.ElementTree as ET
+
 device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
 
+def rotate_around_origin(vec, how_much_rotation):
```
Comment (nigelswenson) [inline gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py]: So long as these aren't being used anywhere yet, im fine with merging

File: gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py [hunk 3]
```diff
@@ -231,237 +270,265 @@ def __init__(self, arm_or_end_effector="hand", frame_skip=15):
...
-            self.random_poses[i]=temp
+            self.random_poses[i] = temp
             random_poses_file.close()
-        self.instance=0#int(np.random.uniform(low=0,high=100))
-
-
+        self.instance = 0  # int(np.random.uniform(low=0,high=100))
 
     # Funtion to get 3D transformation matrix of the palm and get the wrist position and update both those varriables
     def _get_trans_mat_wrist_pose(self):  # WHY MUST YOU HATE ME WHEN I GIVE YOU NOTHING BUT LOVE?
```
Comment (nigelswenson) [inline gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py]: we should probably remove this inline comment

**Review by nigelswenson — CHANGES_REQUESTED** [review]
I left a couple comments in your files. Mainly you need to update the indexing in main_DDPGfD.py because I noticed that one of the indexes you added was different from the comment above listing the order. We should double check that our indexes there line up with the order from kinova_gripper_env.py

File: gym-kinova-gripper/main_DDPGfD.py (hunk 1 above)
Comment (jimzers) [reply gym-kinova-gripper/main_DDPGfD.py:1432]: For measuring finger to object distance, there are two points of sensors for each part of the finger (the two red dots per segment on the finger). I stuck to the dots that were closer to the camera.

A quick breakdown:
2 dots per finger * 3 fingers * 2 segments (distal, proximal) => 12 total spaces for the finger object distance.

We grab 1 dot for each of the segments on the top 2 fingers, totalling to 4 dots.

File: gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py (hunk 2 above)
Comment (jimzers) [reply gym-kinova-gripper/gym_kinova_gripper/envs/kinova_gripper_env.py]: I will remove and keep in a local file until they find use. (Reduces clutter)

**Comment by jimzers** [comment]
@nigelswenson fixed

**Comment by jimzers** [comment]
amen

**Pull request merged**